pages/outputs/[id].jsx

```diff
@@ -0,0 +1,149 @@
+import React from 'react'
+import Head from 'next/head'
+import { Header } from '@oacore/design'
+
+import request from 'api'
+import { fetchMetadata, fetchSimilarTo, fetchCitations } from 'api/outputs'
+import { useStore } from 'store'
+import Meta from 'modules/meta'
+import Template from 'templates/output'
+
+const LOCALE = 'en-GB'
+const CITATION_STYLES = ['apa', 'bibtex']
+
+// Needed for clear development of the data retrieval
+// At the time of development the API was unstable
+const log = (...args) => {
+  if (process.env.NODE_ENV !== 'production')
+    // eslint-disable-next-line no-console
+    console.log(...args)
+}
+
+export async function getStaticProps({ params: routeParams }) {
+  const { id } = routeParams
+
+  const data = {}
+  let revalidate = 60 * 60 * 24 * 7 // seconds, i.e. every week
+
+  try {
+    const rawOutput = await fetchMetadata(id)
+
+    // Strip some properties to optimise network traffic
+    const { fullText: _, ...output } = rawOutput
+
+    const { data: dataProvider } = await request(output.dataProvider)
+    Object.assign(data, {
+      ...output,
+      dataProvider,
+    })
+  } catch (error) {
+    log(error)
+
+    return {
+      props: { error },
+      notFound: true,
+    }
+  }
+
+  try {
+    const doi = data.identifiers?.doi
+    if (!doi) throw new Error('No DOI — no citation')
+
+    const citations = await fetchCitations(doi, {
+      styles: CITATION_STYLES,
+      locale: LOCALE,
+    })
+    data.citations = citations
+  } catch (citationRetrievalError) {
+    log(citationRetrievalError)
+
+    // If any error happens, we pretend, citation could not be generated
+    // or retrieved. This behaviour should be improved because we could generate
+    // citations ourselves using some library.
+    //
+    // Take your time to explore:
+    // - https://citeproc-js.readthedocs.io/en/latest/index.html
+    // - https://citationstyles.org
+    data.citations = []
+  }
+
+  try {
+    const similarOutputs = await fetchSimilarTo(id)
+
+    // Strip some properties to optimise network traffic
+    data.similarOutputs = similarOutputs.map(
+      ({ fullText, ...output }) => output
+    )
+
+    // TODO: Remove this after the API fix
+    data.similarOutputs = data.similarOutputs.map((output) => ({
+      ...output,
+      authors: output.authors.map(({ name }) => name),
+    }))
+
+    if (typeof data.similarOutputs[0].links[0] != 'object') {
+      // eslint-disable-next-line no-console
+      console.log(data.similarOutputs[0].links)
+      throw new Error('Links are not properly formatted at the moment...')
+    }
+  } catch (error) {
+    log(error)
+
+    // If any error happens, we pretend, there were no recommendations
+    // This behaviour could be changed to explicit error reporting but should
+    // be considered deeper.
+    data.similarOutputs = []
+    revalidate = 30
+  }
+
+  return {
+    props: { data },
+    revalidate,
```

Comment:
The `revalidate` parameter means for Next.js time to cache expiry. If the page passes everything without errors, it's one week. If an error happens (i.e. recommender fails), it becomes 30 seconds. Check the logic above.

The timing could be optimised better based on the type of error, i.e. if the recommender fails, it's 30 seconds, but if it's intentionally nothing, it could be a week still. For the citation generation, it could be similar: if there is no DOI, nothing could help but if doi.org returns an error, we can regenerate the page sooner.

It's for static props generation only, if I am not mistaken.